modules/_labs/side-panel/react/lib/SidePanel.tsx

```diff
@@ -0,0 +1,256 @@
+/** @jsx jsx */
+import * as React from 'react';
+import styled from '@emotion/styled';
+import {css, jsx, keyframes, CSSObject} from '@emotion/core';
+import {IconButton, IconButtonProps} from '@workday/canvas-kit-react-button';
+import {spacing, colors, depth} from '@workday/canvas-kit-react-core';
+import {transformationImportIcon} from '@workday/canvas-system-icons-web';
+
+export type SidePanelVariant = 'standard' | 'alternate';
+export type SidePanelTransitionStates = 'collapsed' | 'collapsing' | 'expanded' | 'expanding';
+
+export interface SidePanelProps extends React.HTMLAttributes<HTMLDivElement> {
+  /**
+   * The element the side panel will render as (e.g. 'div').
+   *
+   * @default 'section'
+   */
+  as?: React.ElementType;
+  /**
+   * The width of the component (in `px` if it's a `number`) when it is collapsed.
+   *
+   * @default 64
+   */
+  collapsedWidth?: number | string;
+  /**
+   * If true, sets the expanded state of the side panel
+   * @default true
+   */
+  expanded?: boolean;
+  /**
+   * The width of the component (in `px` if it's a `number`) when it is expanded.
+   *
+   * @default 320
+   */
+  expandedWidth?: number | string;
+  /**
+   * The height of the component (in `px` if it's a `number`).
+   *
+   * @default '100%'
+   */
+  height?: number | string;
+  /**
+   * Which side the side panel is meant to originate from.
+   *
+   * @default 'left'
+   */
+  origin?: 'left' | 'right';
+  /**
+   * The function called when the side panel's `expanded` state changes. States like `'collapsing'` and `'expanding'` are tracked in another callback: `onStateChange`
+   *
+   * @param boolean
+   */
+  onExpandedChange?: (expanded?: boolean) => void;
+  /**
+   * The function called when the side panel is transitioning between states.
+   * Use this to track when the side panel is animating between `'collapsed'`, `'collapsing'`, `'expanded'`, and `'expanding'` states.
+   * This can be particularly helpful if child components need to react specifically to these states.
+   *
+   * @param SidePanelTransitionStates
+   */
+  onStateTransition?: (state?: SidePanelTransitionStates) => void;
+  /**
+   * The style variant of the side panel. 'standard' is with a `soap100` background, no depth. 'alternate' is a `frenchVanilla100` background with a level 3 depth.
+   *
+   * @default 'standard'
+   */
+  variant?: SidePanelVariant;
+}
+
+const createKeyframes = (from: number | string, to: number | string) => {
+  const normalized = {
+    from: typeof from === 'number' ? from + 'px' : from,
+    to: typeof to === 'number' ? to + 'px' : to,
+  };
+
+  return keyframes`
+    from {
+      width: ${normalized.from};
+      min-width: ${normalized.from};
+      max-width: ${normalized.from};
+    } to {
+      width: ${normalized.to};
+      min-width: ${normalized.to};
+      max-width: ${normalized.to};
+    }
+  `;
+};
+
+const containerVariantStyle: {[K in SidePanelVariant]: CSSObject} = {
```

Comment:
I think what I was going for was to make sure that this object only had keys of the type `SidePanelVariant` (I also now know I can do this with the `Record` utility type). So in the future if we add a variant, we can make sure it gets accessed appropriately in the code. Not sure if `as const` really plays into this unless we're going for type inference on this object.

Reply:
The WD component team ran into an issue where `CSSProperties` (which we define as `React.CSSProperties & Emotion.CSSObject`) caused performance issues for Typescript compilation. I was able to observe the issue, but not reproduce it. Digging into the types of `CSSObject`, I found this:

```ts
export interface CSSOthersObjectForCSSObject {
  [propertiesName: string]: CSSInterpolation
}

export interface CSSObject
  extends CSSPropertiesWithMultiValues,
    CSSPseudosForCSSObject,
    CSSOthersObjectForCSSObject {}
```

You can see that `CSSObject` is extending from `CSSOthersObjectForCSSObject` which expands all properties to `string`. This gives you autocomplete, but no type safety. `margin` is a suggestion, but `margon` is a perfectly okay mistype.

We think the typecheck slowdown comes from multiple levels of mapped types. `CSSObject` is about 900 properties which could cause slowdowns.

`as const` is a way to narrow values of objects from widened types like `string` to narrower types like `'absolute'`. So `CSSObject` is useful for autocomplete, but will not save you from a mistyped property name.

I asked the WD component team to create an issue with a reproduceable build that shows the slowdown so that we can directly address it. I have experienced periodic slowdowns in the Typescript language service, but I don't know what exactly is causing it. I wouldn't be surprised that mapped types of large interfaces causes performance issues.

A compromise might be to specify which CSSObject properties we intend to use: